src/main/java/net/datafaker/internal/helper/COWMap.java

```diff
@@ -0,0 +1,93 @@
+package net.datafaker.internal.helper;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+
+public class COWMap<K, V> implements Map<K, V> {
+    private volatile Map<K, V> map;
+    private final Supplier<Map<K, V>> mapSupplier;
+
+    public COWMap(Supplier<Map<K, V>> mapSupplier) {
+        this.mapSupplier = mapSupplier;
+        this.map = mapSupplier.get();
+    }
+
+    @Override
+    public int size() {
+        return map.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return map.isEmpty();
+    }
+
+    @Override
+    public boolean containsKey(Object key) {
+        return map.containsKey(key);
+    }
+
+    @Override
+    public boolean containsValue(Object value) {
+        return map.containsValue(value);
+    }
+
+    @Override
+    public V get(Object key) {
+        return map.get(key);
+    }
+
+    @Override
+    public V put(K key, V value) {
```

Comment:
I think this function should be `synchronized` otherwise it might produce incorrect result, don't you think?
For example 2 threads will access this function at the same time, one of the pair of key-value will be lost...

Reply:
No, that's the main idea of COW approach: it is lock free.
Ok let's consider the case when 2 threads accessing this map.
Yes, in worst case one of the value will be lost. However it is ok. 
We use this as a cache. It means that the lost value will be recalculated and put into map next time. 
So there is no issue.